rand_unix: don't discard entropy bytes from syscall_random() and /dev/*random

mspncp opened the pull request:
Fixes #6978

Don't discard partial reads from syscall_random() and retry instead.

Don't discard partial reads from /dev/*random as well.

_Note: does this need a test? Don't really know how to test this in real life._

paulidale commented:
Would a little more sophistication here make sense?  I can think of two points that might help:

1. Rather than retrying the same entropy source immediately, I think it would make sense to move on to the next source and only retry a source once all have failed to provide enough bytes.  I.e. make the loop around all of the sources rather than one.  This will require changes to our early exit checks that only look to see if some entropy was obtained.
2. Instead of a fixed iteration limit, would it make sense to iterate until no entropy was gather in the loop?  While we are still making progress, keep asking.  If the progress stalls, give up.

Of course, if only one source is defined the first item makes no difference but when multiple are it could.

Thoughts?

paulidale commented:
I'm willing to live without a test.  A test could implement a `getrandom()` function that always returns a single byte but this only works if that is the only configured source.

To manually test it would be possible to drain /dev/random using something like `dd if=/dev/random of=/dev/null iflag=fullblock bs=16` and then run a test where /dev/random is the only source configured.

mspncp commented:
> 1. Rather than retrying the same entropy source immediately, I think it would make sense to move on to the next source...

I considered this particular possibility and discarded it, because the `DEVRANDOM` list contains non-blocking and blocking entropy sources. So I'm not sure if this would be a good idea.

```
#   define DEVRANDOM "/dev/urandom","/dev/random","/dev/srandom"

```

> 2. Instead of a fixed iteration limit, would it make sense to iterate until no entropy was gather in the loop?  While we are still making progress, keep asking.  If the progress stalls, give up.

The main point about my choice was that it's bound to terminate. Your proposal sounds like a good alternative. Or maybe a mixture of our ideas, i.e. reset the retry counter whenever we read more than 0 bytes? 

> To manually test it would be possible to drain /dev/random using something like `dd if=/dev/random of=/dev/null iflag=fullblock bs=16` and then run a test where /dev/random is the only source configured.

Sounds like a good idea. I will give it a try.

> Thoughts?

I'm not in a hurry to merge. We can discuss this without rush.

paulidale commented:
I'd be fine with a making progress and an iteration limit (either an absolute one or one that is reset on progress as you suggest).

What about opening the /dev/random devices with the `O_NONBLOCK` flag so they return rather than blocking?  The error handling would need to check for either `EAGAIN` or `EWOULDBLOCK` (either of which might not be defined and if both are they might or might not have the same value) so it gets a little more complicated.

I agree, let's take the time to do this properly.

mspncp commented:
> What about opening the /dev/random devices with the O_NONBLOCK flag so they return rather than blocking? 

Wouldn't that mean that we would busy loop on /dev/random when it is drained and/or fall through the loop without obtaining enough entropy ? I'm not sure whether this is a good idea.

paulidale commented:
If we add non-blocking I/O for the DEVRANDOM sources, we'd also need the fall through to check other sources -- that was the suggestion you commented about or have I misunderstood something?

We'd still need the iteration limiting of course.

dot-asm commented on crypto/rand/rand_unix.c:
```diff
@@ -441,16 +441,21 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
-        rand_pool_add_end(pool, bytes, 8 * bytes);
-        entropy_available = rand_pool_entropy_available(pool);
+        bytes_needed = rand_pool_bytes_needed(pool, 1 /*entropy_factor*/);
+        while(bytes_needed != 0 && attempts-- > 0) {
```
Style nit. It's `function(arg)`, but `while (condition)`.

mspncp replied:
Thank, I overlooked that one. Fixed in fe675d2bfc31.

dot-asm commented on crypto/rand/rand_unix.c:
```diff
@@ -441,16 +441,21 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
+            buffer = rand_pool_add_begin(pool, bytes_needed);
+            bytes = syscall_random(buffer, bytes_needed);
+            if (bytes > 0) {
+                rand_pool_add_end(pool, (size_t)bytes, 8 * (size_t)bytes);
```
Here is interesting question. Is explicit cast actually required? And if not, are we better off without it? Cast is not actually required because of explicit bytes > 0 check, and implicit conversion would work predictably. But are we better off with it? I'm **not** saying that I have an answer, but I can't help noting that with unsigned multiplication being specified as modulo word-size operation, and signed can turning undefined if multiplicands are large enough, explicit cast renders outcome of multiplication by 8 unconditionally valid. And without it we would give undefined behaviour sanitizer to catch it for us... Does it mean that a check for result of multiplication to be guaranteed appropriate? Again, not that I know the definitive answer, but *formally speaking* yes...

mspncp replied:
> Here is interesting question. Is explicit cast actually required?

I really love your interesting questions. Very often I learn a lot when following the discussions they trigger :-). I must admit that it was a pure reflex that I placed the cast in anticipation of a compiler warning. You were right, there is no warning about sign conversion when I remove the casts. But the reason you gave seems to be incorrect, there is also no warning even if I replace `if (bytes > 0)` by  `if (bytes < 0)`. The true cause seems to be that `-Wsign-conversion` is disabled even when configuring  `--strict-warnings`. Here's my configuration:

```
/usr/bin/perl ./Configure debug-linux-x86_64 threads shared enable-egd enable-crypto-mdebug -g --strict-warnings --prefix=/home/msp/install/openssl
```

I tried to add `-Wsign-conversion` explicitely and quickly gave up, because there were tons of warnings.

mspncp replied:
But putting compilers and sanitizers aside: Yes, I agree with you that because of the if-clause it is safe to omit the cast.

dot-asm replied:
On non-direct, vaguely related note, one can only wonder how did they reason declaring call as one taking size_t and returning int. Of course that large lengths are not something anybody would ever ask for, but declaration asks for question what happens if you ask for 4GB? Would it return 2GB-1? Again, just a side note...

mspncp replied:
@dot-asm please have a look at 23d7c0985115f412b0cc117fafd928f7d575bb7f where I addressed your comment about the return value of syscall_random().

dot-asm replied:
Note that I didn't actually suggest to change the declaration, only wondered what did **they** think when declaring underlying call like that. I mean syscall_random is our function, but it calls system-specific one that accepts size_t and returns int, and remark was about the latter. But now when I look at it in morning light, I realize that it was misdirected, because underlying call doesn't return amount of bytes, but error or success. Duh! This however makes question applicable to syscall_random...

mspncp commented:
@paulidale I am very sceptical about your suggestion to introduce non-blocking I/O for blocking random devices (see below). But independently of my scepticism you are surely leaving the path of a simple bugfix for #6978 and heading towards a redesign of multiple entropy source handling (#4394). As such I would say that it is not an option in the current late beta phase.

You might argument that already my handling of EINTR is more than a bug fix for #6978, i.e, more than just keeping the partially read input. I have to agree, but in this case things seem more simple and predictable. Doing an immediate retry is the normal way of handling EINTR and probably not even a retry counter would be necessary. (Shall I remove it?)

My argument against non-blocking I/O is simple: If you don't want to block, use `/dev/urandom`. In fact, in the case where you have both `/dev/urandom` and `/dev/random` device nodes available, the order in the `DEVRANDOM` list guarantees that the former is used first.  And a read from it will never block except for the early boot stage when the kernel random pool is not initialized yet, in which case it is essential to block and wait.

Looping through the devices in the case of `EAGIN/EWOULDBLOCK` does not make sense for me: It is hard for me to imagine that a non-blocking read from `/dev/urandom` could return `EAGAIN/EWOULDBLOCK` and immediately afterwards a read from `/dev/random` would succeed. Even worse, in the case where only `/dev/random` is available, you would  either busy-loop on `EGAIN/EWOULDBLOCK` errors or drop out of the loop before receiving enough entropy, if the random pool is drained.

mspncp commented:
Travis failure is unrelated:

```
fatal: unable to access 'https://github.com/openssl/openssl.git/': Could not resolve host: github.com
```

paulidale commented:
@mspncp you are right, it's too late in the release cycle for a bigger change.  Let's fix the bug and rework the entropy gathering properly later.

The way things are setup, the complete entropy request will always be fulfilled from /dev/urandom, unless that device doesn't exist.  For FIPS we might need to remove /dev/urandom from the DEVRANDOM list which will make things more interesting.

In the rework, we'll have to consider the case where DEVRANDOM is the only source and blocking makes sense versus having alternative sources where non-blocking does.  A discussion for a future time :)

mspncp commented:
@paulidale, unfortunately your approval overlapped with some changes of mine. In particular I addressed a comment of @dot-asm and I implemented a reset of the `attempt` counter on every successfull read. Please have another look.

mspncp commented on crypto/rand/rand_unix.c:
```diff
@@ -88,6 +86,8 @@ int syscall_random(void *buf, size_t buflen);
     || defined(OPENSSL_SYS_UEFI))
 
+static ssize_t syscall_random(void *buf, size_t buflen);
+
```
Note: The declaration was moved out of sight of the MSVC compiler to fix this [AppVeyor build failure](https://ci.appveyor.com/project/openssl/openssl/build/master.19459/job/j7ufl0mpi9q33ugu).

t-j-h commented on crypto/rand/rand_unix.c:
```diff
@@ -441,16 +441,24 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
-            bytes = bytes_needed;
+    {
+        ssize_t bytes;
+        /* Maximum allowed number of consecutive unsuccessfull attempts */
```
unsuccessful (only one l)

mspncp replied:
Thanks, fixed!

dot-asm commented on crypto/rand/rand_unix.c:
```diff
@@ -217,7 +217,8 @@ static size_t sysctl_random(char *buf, size_t buflen)
 #if   defined(__NetBSD__) && __NetBSD_Version__ < 400000000
-    return 0;
+    errno = ENOSYS;
+    return -1;
```
There is another `return 0` few lines up.

mspncp replied:
done, see b81dc2738bd575543720019ec5ad017517968098

mspncp replied:
it's now in 7722d3d063098af9362eb486468bf3268887b07d.

mspncp commented on crypto/rand/rand_unix.c:
```diff
@@ -256,7 +257,7 @@ int syscall_random(void *buf, size_t buflen)
 
     if (getentropy != NULL)
-        return getentropy(buf, buflen) == 0 ? buflen : 0;
+        return getentropy(buf, buflen) == 0 ? (ssize_t)buflen : -1;
```
done, see 89b93e22bda176dbd15c79fbccdda42a1c493270

dot-asm commented on crypto/rand/rand_unix.c:
```diff
@@ -192,22 +192,30 @@ void rand_pool_keep_random_devices_open(int keep)
+    if (!ossl_assert(buflen <= OSSL_SSIZE_MAX)) {
+        errno = EINVAL;
+        return -1;
+    }
```
Well, this one is actually redundant, because sysctl_random is called from syscall_random, which already performed the check. One can even wonder following. Wouldn't check for buflen <= OSSL_SSIZE_MAX/8 in rand_pool_acquire_entropy be more appropriate? Because it's multiplied by 8 when it's passed to rand_pool_add_end and you don't want that to overflow. Again, I'm not saying that it would, ever, but if we take this path, then it would be only appropriate if we take it all the way. On related note double-check cast in equality test in rand_pool_acquire_entropy...

mspncp replied:
> Well, this one is actually redundant, because sysctl_random is called from syscall_random

Indeed, I considered adding only a comment about it, but then I chose the more defensive solution. I'll remove the redundand check and will also consider your other suggestion about rand_pool_acquire_entropy().

mspncp commented:
Thanks @dot-asm, your mentioning of `rand_pool_acquire_entropy()` was a good hint. A rough estimate of the upper bound for the buffer size yields 2^13. So I removed the asserts and replaced them by comments. I.e., I discarded my last fixups and added aa69fb3c56377fa7d6e87856b535bdc5fd71baff.

mspncp commented:
Rebased branch autosquashing fixups without further changes.

paulidale commented on crypto/rand/rand_unix.c:
```diff
@@ -238,10 +246,20 @@ static size_t sysctl_random(char *buf, size_t buflen)
+     * Note: 'buflen' equals the size of the buffer which is used by the
+     * get_entropy() callback of the RAND_DRBG. It is roughly bounded by
+     *
+     *   2 * DRBG_MINMAX_FACTOR * (RAND_DRBG_STRENGTH : 8) = 2^13
```
`: 8` should that be `/ 8` to keep C syntax?

paulidale replied:
I'd prefer this changed but it isn't essential.

mspncp replied:
Ups, you're right! I'll change it. Seems like I had too many math sessions with my daughter recently ;-).

paulidale replied:
I know the feeling :)

t-j-h commented on crypto/rand/rand_unix.c:
```diff
@@ -468,22 +493,27 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
+                if (bytes > 0) {
+                    rand_pool_add_end(pool, bytes, 8 * bytes);
+                    bytes_needed -= bytes;
+                    attempts = 3; /* reset counter after successfull attempt */
```
One more typo here - successfull -> successful

mspncp commented:
> Please fix the one comment typo 

Done. Puh, that typo was really persistent...

mspncp commented:
Merged, thank you everybody!
